src/crypto.rs

```diff
@@ -72,6 +122,349 @@ pub fn verify_signed_xml<Bytes: AsRef<[u8]>>(
+    for sig_node in signature_nodes.drain(..) {
+        if let Some(sig_root_node) = get_signed_node(&sig_node, doc) {
+            let mut nodes = Vec::new();
+            let mut parent = sig_root_node.get_parent();
+            nodes.push(sig_root_node);
+
+            // mark all children
+            while nodes.len() > 0 {
+                let node = nodes.pop().unwrap();
+                let node_ptr = node.node_ptr() as usize;
+                for child in node.get_child_elements() {
+                    nodes.push(child);
+                }
+                ptr_to_required_node.entry(node_ptr).or_insert(node);
+            }
+
+            // mark the ancestor chain
+            while let Some(p) = parent {
+                let p_ptr = p.node_ptr() as usize;
+                parent = p.get_parent();
+                ptr_to_required_node.entry(p_ptr).or_insert(p);
+            }
+        }
+    }
+    drop(root_elem);
+    for node in ptr_to_required_node.values_mut() {
+        node.set_attribute_ns(ATTRIB_SIGVER, VALUE_SIGVER, ns).unwrap();
+    }
+}
+
+/// Remove all elements that do not contain a signature-verified attribute ([`ATTRIB_SIGVER`] in
+/// the namespace [`XMLNS_SIGVER`]).
+#[cfg(feature = "xmlsec")]
+fn remove_unverified_elements(node: &mut libxml::tree::Node) {
+    // depth-first
+    for mut child in node.get_child_elements() {
+        remove_unverified_elements(&mut child);
+    }
+
+    if node.get_attribute_ns(ATTRIB_SIGVER, XMLNS_SIGVER) != Some(String::from(VALUE_SIGVER)) {
+        // element is unverified; remove it
+        node.unlink_node();
+    }
+}
+
+/// Takes an XML document, parses it, verifies all XML digital signatures against the given
+/// certificates, and returns a derived version of the document where all elements that are not
+/// covered by a digital signature have been removed.
+#[cfg(feature = "xmlsec")]
+pub(crate) fn reduce_xml_to_signed(
+    xml_str: &str,
+    certs: &[openssl::x509::X509],
+) -> Result<String, Error> {
+    let mut xml = XmlParser::default().parse_string(xml_str)?;
+    let mut root_elem = xml.get_root_element().ok_or(Error::XmlMissingRootElement)?;
+
+    // collect ID attribute values and tell libxml about them
+    collect_id_attributes(&mut xml)?;
+
+    // verify each signature
+    let mut signature_nodes = find_signature_nodes(&root_elem);
+    for sig_node in signature_nodes.drain(..) {
+        let mut sig_ctx = XmlSecSignatureContext::new()?;
+        let mut verified = false;
+        for openssl_key in certs {
+            let key_data = openssl_key.to_der()?;
+            let key = XmlSecKey::from_memory(&key_data, XmlSecKeyFormat::CertDer)?;
+            sig_ctx.insert_key(key);
+            verified = sig_ctx.verify_node(&sig_node)?;
+            if verified {
+                break;
+            }
+        }
+
+        if !verified {
+            return Err(Error::InvalidSignature);
+        }
+    }
+    drop(signature_nodes);
```

Comment:
Same issue as above -- it contains references to nodes I'll be modifying in the subsequent calls...

Reply:
This one is less scary since we're not dropping a function argument. I think idiomatic rust just lets things go out of scope instead of manually writing `drop()`, so just wrap this block in curly braces:
```
{
    let mut signature_nodes = find_signature_nodes(&root_elem);
    for sig_node in signature_nodes.drain(..) {
        let mut sig_ctx = XmlSecSignatureContext::new()?;
        let mut verified = false;
        for openssl_key in certs {
            let key_data = openssl_key.to_der()?;
            let key = XmlSecKey::from_memory(&key_data, XmlSecKeyFormat::CertDer)?;
            sig_ctx.insert_key(key);
            verified = sig_ctx.verify_node(&sig_node)?;
            if verified {
                break;
            }
        }

        if !verified {
            return Err(Error::InvalidSignature);
        }
    }
}
```